File: week_1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int main(void)
{
    // Prompt for input
    long cardID = get_long("Number : ");
    //Count lenght
    long forCount = cardID;
    int len = 0;
    while (forCount != 0)
    {
        forCount = forCount / 10;
        len++;
    }
    long afterDigit = cardID;
    int pair = 0;
    int impair = 0;
    // Checksum for card
    for (int i = 0; i < len; i++)
    {
        if (i % 2 == 0)
        {
            pair += afterDigit % 10;
            afterDigit = afterDigit / 10;
        }
        else
        {
            int digitImpair = 0;
            digitImpair = (afterDigit % 10) * 2;
            if (digitImpair >= 10)
            {
                digitImpair = (digitImpair - 10) + 1;
            }
            impair += digitImpair;
            afterDigit = afterDigit / 10;
        }
    }

    // verify the card
    long twoDigit = pow(10, len - 2);
    long oneDigit = pow(10, len - 1);
    bool american = (len == 15) && (cardID / twoDigit == 34 || cardID / twoDigit == 37);
    bool master = (len == 16) && (cardID / twoDigit == 51 || cardID / twoDigit == 52 || cardID / twoDigit == 53
                                  || cardID / twoDigit == 54 || cardID / twoDigit == 55);
    bool visa = (len == 13 || len == 14 || len == 15 || len == 16) && (cardID / oneDigit == 4);
    
    if ((pair + impair) % 10 == 0 && american)
    {
        printf("AMEX\n");
    }
    else if ((pair + impair) % 10 == 0 && visa)
    {
        printf("VISA\n");
    }
    else if ((pair + impair) % 10 == 0 && master)
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}

File: week_6_Python/credit/credit.py
# Get card number
print("Number : ", end='')
card = int(input())

# Get the lenght of
cardLen = len(str(card))
pair = 0
impair = 0
# Checksum the card
afterDigit = card

for i in range(cardLen):
    if i % 2 == 0:
        pair += afterDigit % 10
        afterDigit = int(afterDigit / 10)
    else:
        digitImpair = 0
        digitImpair = (afterDigit % 10) * 2
        if digitImpair >= 10:
            digitImpair = (digitImpair - 10) + 1
        impair += digitImpair
        afterDigit = int(afterDigit / 10)

# Verify the card
amex = (cardLen == 15) and (int(card / pow(10, cardLen - 2)) == 34 or int(card / pow(10, cardLen - 2)) == 37)

master = (cardLen == 16) and (int(card / pow(10, cardLen - 2)) == 51 or int(card / pow(10, cardLen - 2)) == 52
                              or int(card / pow(10, cardLen - 2)) == 53 or int(card / pow(10, cardLen - 2)) == 54 or int(card / pow(10, cardLen - 2)) == 55)
visa = (cardLen == 13 or cardLen == 14 or cardLen == 15 or cardLen == 16) and (int(card / pow(10, cardLen - 1)) == 4)

if (pair + impair) % 10 == 0 and amex:
    print("AMEX")
elif (pair + impair) % 10 == 0 and master:
    print("MASTERCARD")
elif (pair + impair) % 10 == 0 and visa:
    print("VISA")
else:
    print("INVALID")